### PR TITLE
Allow Prefilling Add Server Dialog With HTTP URLs

### PR DESCRIPTION
**Pull request by Kissaki**
The commits in this PR restructure the ConnectDialogEdit code and implement suggesting to the user to fill the dialog form from the clipboard (if clipboard data is available). Implements #3143 

---
Outdated

2017-07-18 01_42_33-add server

TODO:
* [x] Use existing themeable color/style
* [x] Fix/Need layout auto-adjustment on show/hide
* [x] Implement fill suggestion for current connected to server (fallback if no clipboard data)
* [ ] Prevent vertical resizability

@mkrautz @davidebeatrici @hacst what do you think about this new UI element (as per screenshot)?

Also makes sense for prefilling with the current connected to server instead of autofilling like we currently do, right? (Even though we can not have invalid/wrong data there like for HTTP URLs.)

### CODE REVIEW

**Comment by davidebeatrici** [comment]
It's beautiful.

Try to change the background color to grey or cyan and see if it's better.

**Comment by mkrautz** [comment]
I'd like a button for dismissing the notice.
Also, the button should maybe say "Fill" instead?

**Comment by mkrautz** [comment]
We already have a color for such notices.

https://github.com/mumble-voip/mumble-theme/blob/9dadcf597b192effe85576f0ae01791150772a0b/source/Imports/Base%20Theme.scss#L289-L298

Note that it also has a style in our default theme:
https://github.com/mumble-voip/mumble/blob/91ebb8b/src/mumble/Themes.cpp#L158

**Comment by Kissaki** [comment]
Okay,thanks. I’ll polish this then.

**Comment by klemun** [comment]
I think 'You have an URL...' should be 'You have a URL...'. I'm also working on a mock-up of this window in Illustrator just to see how it could look with different colors.

File: src/mumble/Themes.cpp [hunk 1]
```diff
@@ -155,5 +155,5 @@ bool Themes::readStylesheet(const QString &stylesheetFn, QString &stylesheetCont
 }
 
 QString Themes::getDefaultStylesheet() {
-	return QLatin1String(".log-channel{text-decoration:none;}.log-user{text-decoration:none;}p{margin:0;}#qwMacWarning{background-color:#FFFEDC;border-radius:5px;border:1px solid #B5B59E;}");
+	return QLatin1String(".log-channel{text-decoration:none;}.log-user{text-decoration:none;}p{margin:0;}#qwMacWarning,#qwInlineNotice{background-color:#FFFEDC;border-radius:5px;border:1px solid #B5B59E;}");
```
Comment (mkrautz) [inline src/mumble/Themes.cpp]: I guess we need a PR in mumble-theme as well?

**Comment by Kissaki** [comment]
Ready for review.

I did not manage to make the dialog not resize vertically when the notice is hidden. It works when it shows. Before these changes the dialog was also vertically resizable with no use, so at least it's not worse.

Theme changes to dark and lite will have to be committed separately to the theme repo.

image
image
image
image

File: src/mumble/Themes.cpp (hunk 1 above)
Comment (Kissaki) [inline src/mumble/Themes.cpp]: Yes

**Comment by mkrautz** [comment]
I only did a cursory glance. PR looks great. Easy to review with the split commits. Great job.
I will do a thorough review ASAP, but not now at 00:30. :-)

Quick comments on the layout of the notice:

- I think the X button should be "Ignore"
- Can we stack the buttons vertically instead? (row layout?)

**Comment by Kissaki** [comment]
Labeled the buttons with text and icons (dingbats, unicode characters). Also added a newline to the connected server to match the other text and I think it looks better.
image

**Comment by Kissaki** [comment]
Tried to align them right, but that didn’t work… 🤔

File: src/mumble/ConnectDialog.h [hunk 2]
```diff
@@ -214,6 +214,7 @@ class ConnectDialogEdit : public QDialog, protected Ui::ConnectDialogEdit {
 		void on_qcbShowPassword_toggled(bool);
 		void on_qleName_textEdited(const QString&);
 		void on_qleServer_textEdited(const QString&);
+		void showNotice(const QString & text);
```
Comment (mkrautz) [inline src/mumble/ConnectDialog.h]: Coding style: `&` should hug text.

**Comment by mkrautz** [comment]
I have made some adjustments to the layout of the dialog:

image
image

Commit at https://github.com/mkrautz/mumble/commit/9709ed89e62d4d264a32010eadc0f6fb72fbc90b
Feel free to integrate it. (Maybe squash it into the original?)

**Comment by mkrautz** [comment]
Also, I think auto-updating from the clipboard might be surprising...

When I was making the screenshots for my post above, I used Alt+PrintScreen. This caused the notice to disappear because my clipboard was updated with a screenshot.

So, one could argue that it should only refresh its state if the clipboard contains a valid item for the connect dialog.

But even then... I am not sure if automatically syncing from the clipboard should be there. Consider a user that is multitasking. It wouldn't be out of the realm of possibility for a user to open the add server dialog and get distracted. When the user returns to the dialog, the system's clipboard content might now contain something else, or an actual web URL, or whatever. Since the dialog updates automatically, the user might not notice. So I actually think we should back off from this.

**Comment by Kissaki** [comment]
Picked, and fixed format.

**Comment by Kissaki** [comment]
Is the Travis apple build broken?

```
Info: creating stash file /Users/travis/build/mumble-voip/mumble/.qmake.stash
Cannot read /Users/travis/build/mumble-voip/mumble/scripts/scripts.pro: No such file or directory
…
    #include <Ice/SliceChecksumDict.ice>
1 error in preprocessor.
```
Retriggering did not help.

**Comment by mkrautz** [comment]
Yes, I don't know what's wrong yet.

**Comment by mkrautz** [comment]
I guess the paths for Homebrew's Ice changed. Probably due to Ice 3.7.

**Comment by Kissaki** [comment]
Regarding the vertical sizing: Before these changes, the dialog can be resized vertically as well, leading to increasing space between the form rows. So this PR does not make it worse.

After changing `vsizetype="MinimumExpanding"` to `vsizetype="Fixed" on the dialog I would have expected it to be unchangeable vertically, which is not the case.
Only after adding a `sizeConstraint` on the layout (`QLayout::SetMinAndMaxSize`) is it only marginally resizable vertically.
Just now I set vertical to Fixed for every control, and it has the same effect but only with the notice open/displayed. Without it, it still expands.

So rather than change anything here I don't understand, I'd rather keep it to allow vertical resizing - for now, and not touch that in this PR.

**Pull request merged**